typing/ctype.ml

```diff
@@ -1179,7 +1181,7 @@ let rec copy ?partial ?keep_names scope ty =
       match partial with
         None -> assert false
       | Some (free_univars, keep) ->
-          if TypeSet.is_empty (free_univars ty) then
+          if not (is_Tpoly ty) && TypeSet.is_empty (free_univars ty) then
```

Comment:
This was a preextisting bug in `copy`. When `partial` is set the copy is allowed to not copy the whole type -- replacing some bits with unification variables -- which is fine because the type will later on be unified with a full copy of the type. However, it shouldn't stop copying on a `Tpoly` node because that will break the invariant that a node should always be a `Tpoly` in certain positions.